rust-demangle.c

```diff
@@ -732,45 +787,138 @@ static void demangle_dyn_trait(struct rust_demangler *rdm) {
         PRINT(">");
 }
 
-static void demangle_const(struct rust_demangler *rdm) {
+static void demangle_const(struct rust_demangler *rdm, bool in_value) {
     CHECK_OR(!rdm->errored, return );
 
-    if (eat(rdm, 'B')) {
-        size_t backref = parse_integer_62(rdm);
-        if (!rdm->skipping_printing) {
-            size_t old_next = rdm->next;
-            rdm->next = backref;
-            demangle_const(rdm);
-            rdm->next = old_next;
-        }
-        return;
-    }
+    bool opened_brace = false;
 
     char ty_tag = next(rdm);
     switch (ty_tag) {
+    case 'p':
+        PRINT("_");
+        break;
+
     // Unsigned integer types.
     case 'h':
     case 't':
     case 'm':
     case 'y':
     case 'o':
     case 'j':
+        demangle_const_uint(rdm, ty_tag);
+        break;
+
+    case 'a':
+    case 's':
+    case 'l':
+    case 'x':
+    case 'n':
+    case 'i':
+        if (eat(rdm, 'n')) {
+            PRINT("-");
+        }
+        demangle_const_uint(rdm, ty_tag);
+        break;
+
+    case 'b': {
+        uint64_t value = 0;
+        size_t hex_len = 0;
+        while (!eat(rdm, '_')) {
+            value <<= 4;
+
+            char c = next(rdm);
+            if (IS_DIGIT(c))
+                value |= c - '0';
+            else if (c >= 'a' && c <= 'f')
+                value |= 10 + (c - 'a');
+            else
+                ERROR_AND(return );
+            hex_len++;
+        }
+
+        if (value == 0) {
+            PRINT("false");
+        } else if (value == 1) {
+            PRINT("true");
+        } else {
+            ERROR_AND(return );
+        }
+        break;
+    }
```

Comment:
Don't we require `b0_` and `b1_`? Oh, no, we don't, that got changed with `try_parse_uint`.
And `demangle_const_uint` doesn't have a `try_parse_uint` because of how it handles hex fallback, ugh.